src/librustc/ty/mod.rs

```diff
@@ -2142,6 +2142,11 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
 
     pub fn associated_item(self, def_id: DefId) -> AssociatedItem {
         queries::associated_item::get(self, DUMMY_SP, def_id)
+            .expect("not an associated item")
+    }
+
+    pub fn opt_associated_item(self, def_id: DefId) -> Option<AssociatedItem> {
```

Comment:
Can't we avoid introducing queries like these? i.e. actually check what kind of `DefId` you have before calling `associated_item`.